src/platform_linux/platform.c

```diff
@@ -202,6 +202,140 @@ platform_spinlock_destroy(platform_spinlock *lock)
    return CONST_STATUS(ret);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * Batch Read/Write Locks
+ *
+ * These are generic distributed reader/writer locks with an intermediate claim
+ * state. These offer similar semantics to the locks in the clockcache, but in
+ * these locks read locks cannot be held with write locks.
+ *
+ * These locks are allocated in batches of PLATFORM_CACHELINE_SIZE / 2, so that
+ * the write lock and claim bits, as well as the distributed read counters, can
+ * be colocated across the batch.
+ *-----------------------------------------------------------------------------
+ */
+
+void
+platform_batch_rwlock_init(platform_batch_rwlock *lock)
+{
+   ZERO_CONTENTS(lock);
+}
+
+/*
+ *-----------------------------------------------------------------------------
+ * lock/unlock
+ *
+ * Drains and blocks all gets (read locks)
+ *
+ * Caller must hold a claim
+ * Caller cannot hold a get (read lock)
+ *-----------------------------------------------------------------------------
+ */
+
+void
+platform_batch_rwlock_lock(platform_batch_rwlock *lock, uint64 lock_idx)
+{
+   platform_assert(lock->write_lock[lock_idx].claim);
+   debug_only uint8 was_locked =
+      __sync_lock_test_and_set(&lock->write_lock[lock_idx].lock, 1);
+   debug_assert(!was_locked,
+                "platform_batch_rwlock_lock: Attempt to lock a locked page.\n");
+
+   uint64 wait = 1;
+   for (uint64 i = 0; i < MAX_THREADS; i++) {
+      while (lock->read_counter[i][lock_idx] != 0) {
+         platform_sleep(wait);
+         wait = wait > 2048 ? wait : 2 * wait;
+      }
+      wait = 1;
+   }
+}
+
+void
+platform_batch_rwlock_unlock(platform_batch_rwlock *lock, uint64 lock_idx)
+{
+   __sync_lock_release(&lock->write_lock[lock_idx].lock);
+}
+
+/*
+ *-----------------------------------------------------------------------------
+ * try_claim/claim/unlock
+ *
+ * A claim blocks all other claimants (and therefore all other writelocks,
+ * because writelocks are required to hold a claim during the writelock).
+ *
+ * Cannot hold a get (read lock)
+ *
+ * try_claim returns whether the claim succeeded
+ *-----------------------------------------------------------------------------
+ */
+
+bool
+platform_batch_rwlock_try_claim(platform_batch_rwlock *lock, uint64 lock_idx)
+{
+   if (__sync_lock_test_and_set(&lock->write_lock[lock_idx].claim, 1)) {
+      return FALSE;
+   }
+   return TRUE;
+}
+
+void
+platform_batch_rwlock_claim(platform_batch_rwlock *lock, uint64 lock_idx)
+{
+   uint64 wait = 1;
+   while (!platform_batch_rwlock_try_claim(lock, lock_idx)) {
+      platform_sleep(wait);
+      wait = wait > 2048 ? wait : 2 * wait;
```

Comment:
After seeing multiple instances of this line in this file, it occurred to me that it may be useful to package this in an inline function, e.g., `wait = platform_lock_waitfor(wait);`

That way, if we ever change our minds on how the `wait` value is managed, we need to change it in just one place.